### PR TITLE
 [TE] frontend - aaronucsd/Add ember data to TE on Dashboard (Home)

### PR DESCRIPTION
**Pull request by aaronucsd**
screen shot 2018-04-17 at 2 07 14 pm
1. Added the models, adapters, and serilizers for ember data for the anomaly  apis.
2. Added new anomaly service to replace the need of `utils/api/anomaly.js` and `utils/anomaly.js` as the adapters will do those internally.
   This anomaly servie has a built in humanized caching to cache the ember objects meant for display purposes only and not mutate the actual proxy models (records).
3. Added a new query cache service. This will add a new caching layer serivce to save the `union wrapper` data that ember-data do not cache in the store.
Ember data will only cache the entities (resources) only. This service can be used for other client persistant libs ($ajax, redux etc).
4. Updated Dashboard to use the new `api/anomalies/services.js` methods.
5. Made fix for the styles for the metric tables.
6. Added the pagination for the metric tables vs `Show more` link as table will not support that.
7. Fixed the time picker codes for dashboard to work now with the cachhing as well. Reload of the same url query will pull from the cache now and draw instantly.
8. TODO: tests for the new services
9. TODO: Write guideline to use ember data conversion for ThirdEye.
10. TODO: Update the anomaly's resolution option to also use ember data.

### CODE REVIEW

File: thirdeye/thirdeye-frontend/app/serializers/performance.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import DS from 'ember-data';
+
+export default DS.JSONSerializer.extend({
+  /*
+   * @summary normalizing the payload from api response simpler or legacy json that may not support the http://jsonapi.org/ spec
+   */
+  normalizeQueryResponse(store, primaryModelClass, payload) {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/serializers/performance.js]: Can we also document the params in the jsdoc?

Comment (justYves) [inline thirdeye/thirdeye-frontend/app/serializers/performance.js]: Can we also document the params in the jsdoc?

File: thirdeye/thirdeye-frontend/app/serializers/base.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    });
+
+    return {
+      data
+    };
+  }
+  /*
+   * serializing the data to send to the api server
+   */
+  // serialize(snapshot, options) {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/serializers/base.js]: do we still need this?

File: thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import Service from '@ember/service';
+import { inject as service } from '@ember/service';
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js]: Do we need to import both Service and inject ?

File: thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+       cached = this._cache[adapterOptions.cacheKey] = this.get('store').query(type, query, adapterOptions);//add to our dictionary
+    }
+
+    return cached;
+  },
+  /*
+  We can use the `url` as the `cacheKey` in the `query` hook.
+  which would be different with each change in query params.
+  */
+  urlForQueryKey(modelName, query) {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js]: let's document the params in the doc above, also, let's stick to the JS doc comment syntax: 
```
/**
 * This is a JS doc comment
 * @type/@method
 * @param
 * @example
 * @return
 */
```

File: thirdeye/thirdeye-frontend/app/pods/services/api/anomalies/service.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * @type {Ember.Service}
+ * @summary This service provides all the api calls for anomalies related data. An `Ember.Service`
+   is a long-lived Ember object that can be made available in different parts of your application.
+ * @example anomaliesApiService: service('services/api/anomalies');
+ */
+export default Service.extend({
+  store: service('store'),
+  queryCache: service('services/query-cache'),
+  init() {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/services/api/anomalies/service.js]: Do we need this `init()` if we're only calling `this_super()`?

File: thirdeye/thirdeye-frontend/app/pods/components/stats-cards/component.js [hunk 6]
```diff
@@ -12,6 +12,7 @@
  */
 import Component from '@ember/component';
 import { set, get } from '@ember/object';
+import { humanizeChange } from 'thirdeye-frontend/utils/utils';
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/components/stats-cards/component.js]: Is this import being used here?

File: thirdeye/thirdeye-frontend/app/pods/custom/anomalies-table/current-wow/template.hbs [hunk 7]
```diff
@@ -1,4 +1,4 @@
 <p>{{record.current}}/{{record.baseline}}</p>
 <p class="te-anomaly-table__value-label te-anomaly-table__value-label--{{calculate-direction record.change}}">
-  ({{record.humanizedChange}})
+  ({{record.humanizedChangeDisplay}})
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/custom/anomalies-table/current-wow/template.hbs]: nit: Let's make this a template helper if this is a UI only display property

File: thirdeye/thirdeye-frontend/app/serializers/performance.js (hunk 1 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/serializers/performance.js]: Yes! We should. Will do that. Thanks!

File: thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js (hunk 3 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js]: Ah my oversight. Good catch!

File: thirdeye/thirdeye-frontend/app/pods/components/stats-cards/component.js (hunk 6 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/components/stats-cards/component.js]: looks like not. Good catch.

File: thirdeye/thirdeye-frontend/app/pods/services/api/anomalies/service.js (hunk 5 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/services/api/anomalies/service.js]: Nope. You're right. I had something there before but removed it. Let me remove this init now. Thanks!

File: thirdeye/thirdeye-frontend/app/pods/components/stats-cards/component.js (hunk 6 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/components/stats-cards/component.js]: removed

File: thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js (hunk 3 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/services/query-cache/service.js]: Actually yes. One is for ` Service.extend(` and the other is to inject `store: service('store'),`.

File: thirdeye/thirdeye-frontend/app/serializers/performance.js (hunk 1 above)
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/serializers/performance.js]: Actually this is a built in hook. Like `model`, `preModel` hooks in route. But i can explicitly document it as well.

Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/serializers/performance.js]: /**
    * @summary A serializer hook. Normalizing the payload from api response simpler or legacy json that may not support the http://jsonapi.org/ spec
    * @method normalizeQueryResponse
    * @param {DS.Store} store
    * @param {DS.Model} primaryModelClass
    * @param {Object} payload - the data object
    * @return {Obect}
    */

**Comment by aaronucsd** [comment]
Updates with @justYves feedbacks. Thanks!

File: thirdeye/thirdeye-frontend/app/pods/home/route.js [hunk 8]
```diff
@@ -48,38 +51,50 @@ export default Route.extend({
...
+      duration,
+      applications
     });
   },
 
   afterModel(model) {
-    const appName = model.appName || model.applications[0].application;
-    const startDate = Number(model.startDate) || this.get('startDate');
+    // Overrides with params if exists
+    const appName = model.appName || null;//model.applications.get('firstObject.application');
```
Comment (newsummit) [inline thirdeye/thirdeye-frontend/app/pods/home/route.js]: Is this comment a TODO?

File: thirdeye/thirdeye-frontend/app/adapters/base.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  // query: function query(store, type, query) {
+  //   const url = `${this.get('namespace')}/${query.appName}`;
+  //   delete query.appName;//remove from query as not needed for actual request
+  //   return this.ajax(url, 'GET', { data: query });
+  // },
+  // The urlForQuery works like the `query` hook above. Both allow mutating the request url.
+  urlForQuery (query, modelName) {
+    /* The switch allows for adding more model names that needs custom request url */
+    switch(modelName) {
+      case 'performance':
```
Comment (newsummit) [inline thirdeye/thirdeye-frontend/app/adapters/base.js]: Thanks for setting this up: store/models/adapters. Much simpler than what we had before.

File: thirdeye/thirdeye-frontend/app/pods/services/api/anomalies/service.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  async queryAnomaliesByAppName(appName, start) {
+    assert('you must pass appName param as an required argument.', appName);
+    assert('you must pass start param as an required argument.', start);
+
+    const queryCache = this.get('queryCache');
+    const modelName = 'anomalies';
+    const query = { application: appName, start };
+    const anomalies = await queryCache.query(modelName, query, { reload: false, cacheKey: queryCache.urlForQueryKey(modelName, query) });
+    return anomalies;
+    //const url = anomalyApiUrls.getAnomaliesByAppNameUrl(appName, startTime);//TODO: remove from the utils/api/anomaly.js - lohuynh
```
Comment (newsummit) [inline thirdeye/thirdeye-frontend/app/pods/services/api/anomalies/service.js]: I see. So this new architecture renders our API utils obsolete. One of the reasons for having those was to have a centralized inventory of all APIs used. What do you recommend for that?

**Comment by aaronucsd** [comment]
Updated with humanized caching to avoid mutating the actual ember proxy models (records from store).

**Pull request merged**